lib/Relation.php

```diff
@@ -0,0 +1,557 @@
+     *
+     * $users = User::readonly()
+     * $users->first()->save()  // throws exception ActiveRecord::ReadOnlyRecord: User is marked as readonly
+     */
+    public function readonly(bool $readonly): Relation
+    {
+        $this->options['readonly'] = $readonly;
+
+        return $this;
+    }
+
+    /**
+     * Find by id - This can either be a specific id (1),
+     * a list of ids (1, 5, 6), or an array of ids ([5, 6, 10]).
+     *
+     * If one or more records cannot be found for the requested ids,
+     * then ActiveRecord::RecordNotFound will be raised.
+     *
+     * If the primary key is an integer, find by id coerces its arguments by using to_i.
+     *
+     * Person.find(1)          # returns the object for ID = 1
+     * Person.find("1")        # returns the object for ID = 1
+     * Person.find("31-sarah") # returns the object for ID = 31
+     * Person.find(1, 2, 6)    # returns an array for objects with IDs in (1, 2, 6)
+     * Person.find([7, 17])    # returns an array for objects with IDs in (7, 17)
+     * Person.find([1])        # returns an array for the object with ID = 1
+     *
+     * Person.where("administrator = 1").order("created_on DESC").find(1)
+     *
+     * Person.find(-1)          # throws RecordNotFound
+     *
+     * Person.find()           # throws ValidationsArgumentError as there's no where statement to execute
+     *
+     * @throws RecordNotFound if any of the records cannot be found
+     *
+     * @return Model|array<Model>
+     */
+    public function find()
+    {
+        $args = func_get_args();
+        $num_args = count($args);
+        $class = $this->className;
+
+        if (0 === $num_args) {
+            throw new ValidationsArgumentError('find requires at least one argument');
+        }
+
+        // find by pk
+        if (1 === $num_args) {
+            $args = $args[0];
+        }
+
+        $single = !is_array($args) || !array_is_list($args);
+
+        $options = $this->options;
+        $options['conditions'] ??= [];
+        $options['conditions'][] = $this->pk_conditions($args);
+        $options['mapped_names'] = $this->alias_attribute;
+
+        $list = $this->table()->find($options);
+        if (is_array($args) && count($list) != count($args)) {
+            throw new RecordNotFound('found ' . count($list) . ', but was looking for ' . count($args));
+        }
+
+        return $single ? ($list[0] ?? throw new RecordNotFound('tbd')) : $list;
+    }
+
+    /**
+     * Returns a record (or N records if a parameter is supplied) without any implied
+     * order. The order will depend on the database implementation.
+     * If an order is supplied it will be respected.
+     *
+     *   Person::take() # returns an object fetched by SELECT * FROM people LIMIT 1
+     *   Person::take(5) # returns 5 objects fetched by SELECT * FROM people LIMIT 5
+     *   Person::where(["name LIKE '%?'", name])->take()
+     */
```

Comment:
Why return mixed?
Shouldn't it be a Model or Model[] or null?

Reply:
Nice catch b81bea4